File: Math/2235_Add_Two_Integers.java
/*
2235.Add Two Integers
Given two integers num1 and num2, return the sum of the two integers.

Example 1: Input: num1 = 12, num2 = 5 Output: 17
Difficulty : easy

Approach : Bit Manipulation.
1. Use bitwise operations to add two integers without using the '+' operator.
2. The idea is to use the XOR operation to add the bits of num1 and num2 without carrying, and the AND operation followed by a left shift to calculate the carry.
3. Repeat the process until there is no carry left (i.e., num2 becomes
4. The final value of num1 will be the sum of the two integers.

Time Complexity: O(log(max(num1, num2))) since we are processing each bit of the numbers.
Space Complexity: O(1) since we are using a constant amount of space for the calculations.

*/

class Solution {
    public int sum(int num1, int num2) {
        // return num1 + num2;
        while (num2 != 0) {
            int carry = (num1 & num2) << 1;
            num1 = num1 ^ num2;
            num2 = carry;
        }

        return num1;
    }
}

File: Math/2235_Add_Two_Integers.py
"""
2235.Add Two Integers
Given two integers num1 and num2, return the sum of the two integers.

Example 1: Input: num1 = 12, num2 = 5 Output: 17

Difficulty : easy

Approach : Bit Manipulation
1. Use bitwise AND to find the carry bits.
2. Use bitwise XOR to find the sum bits.
3. Shift the carry bits to the left by 1 to add them to the sum bits.
4. Repeat the process until there are no carry bits left.

Time Complexity : O(1)
Space Complexity : O(1)

"""


class Solution:
    def sum(self, num1: int, num2: int) -> int:
        # return num1 + num2
        mask = 0xFFFFFFFF
        max_int = 0x7FFFFFFF

        while num2 != 0:
            carry = ((num1 & num2) << 1) & mask
            num1 = (num1 ^ num2) & mask
            num2 = carry

        return num1 if num1 <= max_int else ~(num1 ^ mask)